### PR TITLE
Snapshot api #79

### PR DESCRIPTION
**Pull request by prateekpandey14**
**Feature Request:**

**Why this Change Required ??**
This commit will add snapshot API struct as well as the API handlers
which is going to use in M-apiserver to perform CRUD operations with 
respect to Snapshot.

**How does this change address the issue ?**
This will be initial commit which add the snapshot API struct and handlers.
issue : openebs/maya#79

### CODE REVIEW

File: volumes/provisioner/jiva/snapshot_list.go [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+	}
+
+	first := true
+	//snapshots := []string{}
+	for _, r := range replicas {
+		if r.Mode != "RW" {
+			continue
+		}
+
+		/*	if first {
```
Comment (AmitKumarDas) [inline volumes/provisioner/jiva/snapshot_list.go]: remove this

File: volumes/provisioner/jiva/snapshot_list.go [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+				s = strings.TrimSuffix(strings.TrimPrefix(s, "volume-snap-"), ".img")
+				fmt.Fprintf(tw, format, s)
+			}
+			tw.Flush()
+		*/
+
+		//for _, r := range replicas {
+		if first {
+			first = false
+			snapdisk, err := getData(r.Address)
```
Comment (AmitKumarDas) [inline volumes/provisioner/jiva/snapshot_list.go]: explain what is happening here

File: volumes/provisioner/jiva/snapshot_list.go [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+			if err != err {
+				return err
+			}
+			out := make([]string, len(snapdisk)+1)
+
+			out[0] = "Name|Created At|Size"
```
Comment (AmitKumarDas) [inline volumes/provisioner/jiva/snapshot_list.go]: Use templating

File: volumes/provisioner/jiva/snapshot_list.go [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+
+func (c *ControllerClient) ListReplicas(path string) ([]Replica, error) {
+	var resp ReplicaCollection
+
+	err := c.get(path+"/replicas", &resp)
+
+	return resp.Data, err
+}
+
+func getChain(address string) ([]string, error) {
```
Comment (AmitKumarDas) [inline volumes/provisioner/jiva/snapshot_list.go]: Explain what is this all about ?

File: volumes/provisioner/jiva/snapshot_list.go [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+	}
+
+	r, err := repClient.GetReplica()
+	if err != nil {
+		return nil, err
+	}
+
+	return r.Chain, err
+}
+func getData(address string) (map[string]command.DiskInfo, error) {
```
Comment (AmitKumarDas) [inline volumes/provisioner/jiva/snapshot_list.go]: Explain this func

File: volumes/provisioner/jiva/snapshot_list.go [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+
+func (c *ReplicaClient) GetReplica() (InfoReplica, error) {
+	var replica InfoReplica
+
+	err := c.get(c.Address+"/replicas/1", &replica)
+
+	return replica, err
+}
+
+func (c *ReplicaClient) get(url string, obj interface{}) error {
```
Comment (AmitKumarDas) [inline volumes/provisioner/jiva/snapshot_list.go]: This implementation is redundant & worth fit to be generic.
Do have a look at k8s.90/client-go/rest/client

Build some generic http client stuff by referring to above. 
Or perhaps use a generic http client lib e.g. goa

File: volumes/provisioner/jiva/snapshot_list.go [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+	resp, err := c.httpClient.Get(url)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	return json.NewDecoder(resp.Body).Decode(obj)
+}
+
+func (c *ReplicaClient) post(path string, req, resp interface{}) error {
```
Comment (AmitKumarDas) [inline volumes/provisioner/jiva/snapshot_list.go]: same comment as above

File: volumes/provisioner/jiva/snapshot_revert.go [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+func (c *ControllerClient) RevertSnapshot(volname string, snapshot string) error {
+
+	annotations, err := command.GetVolumeSpec(volname)
+	if err != nil || annotations == nil {
+
+		return err
+	}
+
+	if annotations.ControllerStatus != "Running" {
+		fmt.Println("Volume not reachable")
```
Comment (AmitKumarDas) [inline volumes/provisioner/jiva/snapshot_revert.go]: Remove this

File: volumes/provisioner/jiva/snapshot_revert.go [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+
+func (c *ControllerClient) RevertSnapshot(volname string, snapshot string) error {
+
+	annotations, err := command.GetVolumeSpec(volname)
+	if err != nil || annotations == nil {
+
+		return err
+	}
+
+	if annotations.ControllerStatus != "Running" {
```
Comment (AmitKumarDas) [inline volumes/provisioner/jiva/snapshot_revert.go]: Use typed constants

File: volumes/provisioner/jiva/snapshot_revert.go [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+		return err
+	}
+
+	//var c *ControllerClient
+	volume, err := command.GetVolume(controller.Address)
+	if err != nil {
+		return err
+	}
+
+	url := controller.Address + "/volumes/" + volume.Id + "?action=revert"
```
Comment (AmitKumarDas) [inline volumes/provisioner/jiva/snapshot_revert.go]: No hard codings

File: command/snapshot_create.go [hunk 11]
```diff
@@ -161,7 +161,7 @@ func (c *ControllerClient) do(method, path string, req, resp interface{}) error
 	bodyType := "application/json"
 	url := path
 	if !strings.HasPrefix(url, "http") {
-		url = c.address + path
+		url = c.Address + path
 
 	}
 	fmt.Println("Do url:", url)
```
[the remark this answers is not in the document]
Comment (prateekpandey14) [reply command/snapshot_create.go]: This was for reference ,removed now.

File: volumes/provisioner/jiva/model.go [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+	Data []Replica `json:"data"`
+}
+
+type MarkDiskAsRemovedInput struct {
+	client.Resource
+	Name string `json:"name"`
+}
+
+// ReplicaClient is Client structure
+type ReplicaClient struct {
```
[the remark this answers is not in the document]
Comment (prateekpandey14) [reply volumes/provisioner/jiva/model.go]: this structure will go off from the volumes_stats.go file once we refactor this to proper 
client pkg and its types.

File: volumes/provisioner/jiva/snapshot_list.go (hunk 6 above)
Comment (prateekpandey14) [reply volumes/provisioner/jiva/snapshot_list.go]: These methods are more specific to Replicas right now.

File: volumes/provisioner/jiva/snapshot_revert.go (hunk 9 above)
Comment (prateekpandey14) [reply volumes/provisioner/jiva/snapshot_revert.go]: I will do this, this should be typed constant. Thanks

File: command/snapshot_create.go [hunk 13]
```diff
@@ -126,12 +126,12 @@ func Snapshot(volname string, snapname string, labels map[string]string) (string
...
 	}
 
-	volume, err := GetVolume(controller.address)
+	volume, err := GetVolume(controller.Address)
 	if err != nil {
 		return "", err
 	}
 
-	url := controller.address + "/volumes/" + volume.Id + "?action=snapshot"
+	url := controller.Address + "/volumes/" + volume.Id + "?action=snapshot"
```
[the remark this answers is not in the document]
Comment (prateekpandey14) [reply command/snapshot_create.go]: I will make it typed contant.

File: command/stats.go [hunk 14]
```diff
@@ -122,3 +122,32 @@ func GetVolAnnotations(volName string) (*Annotations, error) {
 	}
 	return &annotations, nil
 }
+
+func GetVolumeSpec(volName string) (*Annotations, error) {
+	var volume Volume
+	var annotations Annotations
+
+	for key, value := range volume.Metadata.Annotations.(map[string]interface{}) {
```
[the remark this answers is not in the document]
Comment (prateekpandey14) [reply command/stats.go]: This requires some more work for proper struct based, but i will do that.

[the remark this answers is not in the document]
Comment (prateekpandey14) [reply command/stats.go]: Needs some code changes at M-apiserver side

File: volumes/provisioner/jiva/snapshot_create.go [hunk 15]
```diff
@@ -0,0 +1,138 @@
...
+
+	if httpResp.StatusCode >= 300 {
+		content, _ := ioutil.ReadAll(httpResp.Body)
+		return fmt.Errorf("Bad response: %d %s: %s", httpResp.StatusCode, httpResp.Status, content)
+	}
+
+	if resp == nil {
+		return nil
+	}
+	return json.NewDecoder(httpResp.Body).Decode(resp)
```
[the remark this answers is not in the document]
Comment (prateekpandey14) [reply volumes/provisioner/jiva/snapshot_create.go]: Thanks, i will make it more generic, will handle with specific case at runtime.

File: volumes/provisioner/jiva/snapshot_list.go (hunk 1 above)
Comment (prateekpandey14) [reply volumes/provisioner/jiva/snapshot_list.go]: did

File: volumes/provisioner/jiva/snapshot_list.go (hunk 3 above)
Comment (prateekpandey14) [reply volumes/provisioner/jiva/snapshot_list.go]: This is, to result the output in tabular form which is more similar to kubectl way of doing.
anyways will try with template too and see.

**Pull request merged**